src/databricks/labs/blueprint/paths.py

```diff
@@ -235,66 +276,320 @@ class WorkspacePath(Path):
+                        f"not {type(path).__name__!r}"
+                    )
+                    raise TypeError(msg)
+                raw_paths.append(path)
+
+        # Normalise the paths that we have.
+        root, path_parts = self._parse_and_normalize(raw_paths)
+
+        self._drv = ""
+        self._root = root
+        self._path_parts = path_parts
+        self._ws = ws
 
-    def _format_parsed_parts(self, drv, root, parts):  # pylint: disable=arguments-differ
-        # instance method adapted from pathlib.Path
-        if drv or root:
-            return drv + root + self._flavour.join(parts[1:])
-        return self._flavour.join(parts)
+    @classmethod
+    def _parse_and_normalize(cls, parts: list[str]) -> tuple[str, tuple[str, ...]]:
+        """Parse and normalize a list of path components.
+
+        Args:
+            parts: a list of path components to parse and normalize.
+        Returns:
+            A tuple containing:
+              - The normalized drive (always '')
+              - The normalized root for this path, or '' if there isn't any.
+              - The normalized path components, if any, (relative) to the root.
+        """
+        match parts:
+            case []:
+                path = ""
+            case [part]:
+                path = part
+            case [*parts]:
+                path = cls.parser.join(*parts)
+        if path:
+            root, rel = cls._splitroot(path, sep=cls.parser.sep)
+            # No need to split drv because we don't support it.
+            parsed = tuple(str(x) for x in rel.split(cls.parser.sep) if x and x != ".")
+        else:
+            root, parsed = "", ()
+        return root, parsed
+
+    @classmethod
+    def _splitroot(cls, part: str, sep: str) -> tuple[str, str]:
+        # Based on the upstream implementation, with the '//'-specific bit elided because we don't need to
+        # bother with Posix semantics.
+        if part and part[0] == sep:
+            root, path = sep, part.lstrip(sep)
+        else:
+            root, path = "", part
+        return root, path
+
+    def __reduce__(self) -> NoReturn:
+        # Cannot support pickling because we can't pickle the workspace client.
+        msg = "Pickling Workspace paths is not supported."
+        raise NotImplementedError(msg)
+
+    def __fspath__(self):
+        # Cannot support this: Workspace objects aren't accessible via the filesystem.
+        msg = f"Workspace paths are not path-like: {self}"
+        raise NotImplementedError(msg)
+
+    def as_posix(self):
+        return str(self)
+
+    def __str__(self):
+        try:
+            return self._str
+        except AttributeError:
+            self._str = (self._root + self.parser.sep.join(self._path_parts)) or "."
+            return self._str
+
+    def __bytes__(self):
+        # Super implementations are fine.
+        return super(self).__bytes__()
```

Comment:
won't it load the super implementation?

Reply:
There's a different bug here: it should be `super()` instead of `super(self)`.

The superclass will always be loaded, it's really about whether it will invoked and whether this is safe. There are a few categories:
 - _The implementation is in terms of other public APIs_, for all the python versions that we support. These I've marked with the comment you see here. These both act as progress markers to indicate that the implementations have been checked, and when I run the tests with coverage I know I've got a test case that verifies it's fine for all the python versions that we support.
 - _The implementation is in terms of internal `_protected` properties_, which all changed between python 3.11 and 3.12. For these I need an independent implementation, which I'm mainly basing on stripped versions of the default implementation from the (new) abstract base class being introduced in 3.13.
 - _A hybrid of the above._ This I treat like the previous item.

My current strategy is to enumerate the public API (including Dunder methods like this), mark each method, and implement the tests to verify it works across the versions.